File: CPP/Arrays/D10_KadaneAlgorithm.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include "ArrayUtils.h"
using namespace std;

int maxSubarraySum(vector<int> &nums) {
      int n = nums.size();

      if (n == 0) {
            return -1;
      }
      int maxSum = nums[0];
      int currentSum = nums[0];

      for (int i = 1 ; i < n ; i++) {
            currentSum = max(currentSum, currentSum + nums[i]);
            maxSum = max(currentSum, maxSum);
      }

      return maxSum;
}

int main() {
      vector<int> nums1 = {2, 3, -8, 7, -1, 2, 3};
      vector<int> nums2 = {-2, -4};
      vector<int> nums3 = {5, 4, 1, 7, 8};

      cout << maxSubarraySum(nums1) << endl ;
      cout << maxSubarraySum(nums2) << endl ;
      cout << maxSubarraySum(nums3) << endl ;

      return 0;
}     

File: Python/Arrays/D10_KadaneAlgorithm.py
def max_subarray_sum(nums):
    if not nums:
        return -1

    max_sum = current_sum = nums[0]

    for num in nums[1:]:
        current_sum = max(num, current_sum + num)
        max_sum = max(max_sum, current_sum)

    return max_sum

print(max_subarray_sum([2, 3, -8, 7, -1, 2, 3]))  
print(max_subarray_sum([-2, -4]))               
print(max_subarray_sum([5, 4, 1, 7, 8]))         